File: python_files/raylib/single_file_programs/pong.py
import random

import pyray as p

SCREEN_WIDTH: int = 1280
SCREEN_HEIGHT: int = 720


class Ball:
    def reset(self) -> None:
        self.center: p.Vector2 = p.Vector2(
            p.get_screen_width() / 2.0, p.get_screen_height() / 2.0
        )
        self.speed: float = 400.0
        self.direction: p.Vector2 = p.Vector2(
            random.choice([-1, 1]), random.choice([-1, 1])
        )
        self.active: bool = False

    def __init__(self) -> None:
        self.radius: float = 10.0
        self.color: p.Color = p.RED
        self.speed_increment: float = 10.0

        self.reset()

    def update(self, dt: float) -> None:
        # activate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = True

        if self.active:
            self.direction = p.vector2_normalize(self.direction)

            # move
            self.center.x += self.direction.x * self.speed * dt
            self.center.y += self.direction.y * self.speed * dt

            # set bounds
            if self.center.y < self.radius:
                self.center.y = self.radius
                self.direction.y *= -1

            if self.center.y > p.get_screen_height() - self.radius:
                self.center.y = p.get_screen_height() - self.radius
                self.direction.y *= -1

    def draw(self) -> None:
        p.draw_circle_v(self.center, self.radius, self.color)


class Paddle:
    def reset(self) -> None:
        self.rect.y = (p.get_screen_height() - self.rect.height) / 2.0

    def __init__(self, x: float) -> None:
        self.rect: p.Rectangle = p.Rectangle(x, 0, 15.0, 100.0)
        self.color: p.Color = p.BLACK
        self.speed: float = 300.0

        self.reset()

    def update_player(self, dt: float) -> None:
        if p.is_key_down(p.KeyboardKey.KEY_W) and self.rect.y > 0:
            self.rect.y -= self.speed * dt
        if (
            p.is_key_down(p.KeyboardKey.KEY_S)
            and self.rect.y < p.get_screen_height() - self.rect.height
        ):
            self.rect.y += self.speed * dt

    def update_ai(self, dt: float, ball_y: float) -> None:
        if ball_y < self.rect.y + self.rect.height / 2.0 and self.rect.y > 0:
            self.rect.y -= self.speed * dt
        if (
            ball_y > self.rect.y + self.rect.height / 2.0
            and self.rect.y < p.get_screen_height() - self.rect.height
        ):
            self.rect.y += self.speed * dt

    def draw(self) -> None:
        p.draw_rectangle_rec(self.rect, self.color)


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, "Pong")

    score_left: int = 0
    score_right: int = 0

    ball: Ball = Ball()
    player: Paddle = Paddle(10.0)
    ai: Paddle = Paddle(p.get_screen_width() - player.rect.width - 10.0)

    while not p.window_should_close():
        dt: float = p.get_frame_time()

        ball.update(dt)
        player.update_player(dt)
        ai.update_ai(dt, ball.center.y)

        # collision
        if p.check_collision_circle_rec(ball.center, ball.radius, player.rect):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment * dt

        if p.check_collision_circle_rec(ball.center, ball.radius, ai.rect):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment * dt

        # update scores
        if ball.center.x < -ball.radius:
            score_right += 1
            ball.reset()
            player.reset()
            ai.reset()

        if ball.center.x > p.get_screen_width() + ball.radius:
            score_left += 1
            ball.reset()
            player.reset()
            ai.reset()

        p.begin_drawing()
        p.clear_background(p.RAYWHITE)

        # draw scores
        p.draw_text(str(score_left), p.get_screen_width() // 4, 30, 30, p.BLACK)
        p.draw_text(
            str(score_right),
            p.get_screen_width() - p.get_screen_width() // 4,
            30,
            30,
            p.BLACK,
        )

        ball.draw()
        player.draw()
        ai.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/single_file_programs/pong.cpp
#include "center_and_draw_text.h"

#include <raylib.h>
#include <raymath.h>

static const int SCREEN_WIDTH        = 800;
static const int SCREEN_HEIGHT       = 600;
static const char *SCREEN_TITLE      = "Pong";
static const Color SCREEN_BACKGROUND = RAYWHITE;

static const int PLACEMENT_OFFSET = 10.0f;

class Ball {
  public:
    float radius;
    Vector2 center;
    float speed;
    float speed_increment;
    Vector2 direction;
    Color color;
    bool active;

    Ball() : radius(10.0f), speed_increment(10.0f), color(RED) { reset(); }

    void reset(void)
    {
        center    = (Vector2){SCREEN_WIDTH / 2.0f, SCREEN_HEIGHT / 2.0f};
        speed     = 400.0f;
        direction = (Vector2){
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
        };
        active = false;
    }

    void draw(void) const
    {
        DrawCircleV(center, radius, color);

        // draw activation text
        if (!active) {
            center_and_draw_text(
                "press SPACE to begin",
                (Rectangle){0, 0, SCREEN_WIDTH, SCREEN_HEIGHT + 200},
                GetFontDefault(), 20.0f);
        }
    }

    void update(float dt)
    {
        // activate ball
        if (IsKeyPressed(KEY_SPACE)) {
            active = true;
        }

        if (active) {
            direction = Vector2Normalize(direction);

            // move
            center.x += direction.x * speed * dt;
            center.y += direction.y * speed * dt;

            // bounds
            if (center.y < radius || center.y > SCREEN_HEIGHT - radius) {
                direction.y *= -1;
            }
        }
    }
};

class Paddle {
  public:
    Rectangle rect;
    float speed;
    Color color;

    Paddle(float x)
        : rect((Rectangle){x, 0, 10.0f, 100.0f}), speed(300.0f), color(BLACK)
    {
        // set rect.y
        reset();
    }

    void reset(void) { rect.y = (SCREEN_HEIGHT - rect.height) / 2.0f; }

    void draw(void) const { DrawRectangleRec(rect, color); }

    void update_player(float dt)
    {
        if (IsKeyDown(KEY_W) && rect.y > 0) {
            rect.y -= speed * dt;
        }
        if (IsKeyDown(KEY_S) && rect.y < SCREEN_HEIGHT - rect.height) {
            rect.y += speed * dt;
        }
    }

    void update_ai(float dt, float ball_y)
    {
        if (ball_y < rect.y + rect.height / 2.0f && rect.y > 0) {
            rect.y -= speed * dt;
        }
        if (ball_y > rect.y + rect.height / 2.0f &&
            rect.y < SCREEN_HEIGHT - rect.height) {
            rect.y += speed * dt;
        }
    }
};

class GameManager {
  public:
    int score_left;
    int score_right;

    Ball ball;
    Paddle player;
    Paddle ai;

    GameManager()
        : score_left(0), score_right(0), ball(), player(PLACEMENT_OFFSET),
          ai(SCREEN_WIDTH - player.rect.width - PLACEMENT_OFFSET)
    {
    }

    void draw(void) const
    {
        // draw scores
        DrawText(TextFormat("%d", score_left), 200, 30, 30, BLACK);
        DrawText(TextFormat("%d", score_right), SCREEN_WIDTH - 200, 30, 30,
                 BLACK);

        ball.draw();
        player.draw();
        ai.draw();
    }

    void handle_collision(float dt)
    {
        // ball collision paddles
        if (CheckCollisionCircleRec(ball.center, ball.radius, player.rect)) {
            ball.direction.x *= -1;
            ball.speed += ball.speed_increment * dt;
        }
        if (CheckCollisionCircleRec(ball.center, ball.radius, ai.rect)) {
            ball.direction.x *= -1;
            ball.speed += ball.speed_increment * dt;
        }
    }

    void reset_ball_and_paddles(void)
    {
        ball.reset();
        player.reset();
        ai.reset();
    }

    void update_scores(void)
    {
        if (ball.center.x < -ball.radius) {
            score_right++;
            reset_ball_and_paddles();
        }
        if (ball.center.x > SCREEN_WIDTH - ball.radius) {
            score_left++;
            reset_ball_and_paddles();
        }
    }

    void update(float dt)
    {
        ball.update(dt);
        player.update_player(dt);
        ai.update_ai(dt, ball.center.y);

        handle_collision(dt);
        update_scores();
    }
};

int main(void)
{
    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    GameManager game_manager;

    while (!WindowShouldClose()) {

        float dt = GetFrameTime();

        game_manager.update(dt);

        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        game_manager.draw();

        EndDrawing();
    }

    CloseWindow();

    return 0;
}
